libs/ballot-interpreter/src/hmpb-rust/timing_marks.rs

```diff
@@ -443,48 +459,130 @@ pub fn find_partial_timing_marks_from_candidate_rects(
     rects: &[Rect],
     debug: &ImageDebugWriter,
 ) -> Option<Partial> {
+    fn average<T: IntoIterator<Item = PixelPosition>>(values: T) -> PixelPosition {
+        let mut sum = 0;
+        let mut count = 0;
+        for value in values {
+            sum += value;
+            count += 1;
+        }
+        if count == 0 {
+            0
+        } else {
+            sum / count
+        }
+    }
+
     let half_height = (geometry.canvas_size.height / 2) as PixelPosition;
     let half_width = (geometry.canvas_size.width / 2) as PixelPosition;
     let top_half_rects = rects
         .iter()
         .filter(|r| r.top() < half_height)
         .copied()
-        .collect::<Vec<Rect>>();
+        .collect::<Vec<_>>();
     let bottom_half_rects = rects
         .iter()
         .filter(|r| r.top() >= half_height)
         .copied()
-        .collect::<Vec<Rect>>();
+        .collect::<Vec<_>>();
     let left_half_rects = rects
         .iter()
         .filter(|r| r.left() < half_width)
         .copied()
-        .collect::<Vec<Rect>>();
+        .collect::<Vec<_>>();
     let right_half_rects = rects
         .iter()
         .filter(|r| r.left() >= half_width)
         .copied()
-        .collect::<Vec<Rect>>();
-    let mut top_line = find_largest_subset_intersecting_line(
-        &top_half_rects,
-        HORIZONTAL_ANGLE,
-        MAX_BEST_FIT_LINE_ERROR,
-    );
-    let mut bottom_line = find_largest_subset_intersecting_line(
+        .collect::<Vec<_>>();
+
+    let cmp_top_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
+        match a.len().cmp(&b.len()) {
+            std::cmp::Ordering::Equal => {
+                // if the counts are equal, sort by top opposite (where we want the smallest value)
+                average(b.iter().map(|r| r.top())).cmp(&average(a.iter().map(|r| r.top())))
+            }
+            cmp => cmp,
+        }
+    };
+
+    let cmp_bottom_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
+        match a.len().cmp(&b.len()) {
+            std::cmp::Ordering::Equal => {
+                // if the counts are equal, sort by bottom (where we want the largest value)
+                average(a.iter().map(|r| r.bottom())).cmp(&average(b.iter().map(|r| r.bottom())))
+            }
+            cmp => cmp,
+        }
+    };
+
+    let cmp_left_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
```

Comment:
This needs to be a closure because functions cannot capture surrounding variables (i.e. `geometry`). I made all the `cmp_*` functions closures just for consistency even though the above two could be regular `fn`s.

Reply:
I think they'd be able to access `average` even if they were `fn`s.